src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferDictionaryTryAddValueOverGuardedAddAnalyzer.cs

```diff
@@ -0,0 +1,149 @@
+// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the MIT license.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using Analyzer.Utilities;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+using static Microsoft.NetCore.Analyzers.MicrosoftNetCoreAnalyzersResources;
+
+namespace Microsoft.NetCore.Analyzers.Performance
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    public sealed class PreferDictionaryTryAddValueOverGuardedAddAnalyzer : DiagnosticAnalyzer
+    {
+        internal const string RuleId = "CA1863";
+
+        private const string ContainsKey = nameof(IDictionary<dynamic, dynamic>.ContainsKey);
+        private const string Add = nameof(IDictionary<dynamic, dynamic>.Add);
+
+        private static readonly LocalizableString s_localizableTitle = CreateLocalizableResourceString(nameof(PreferDictionaryTryAddValueTitle));
+        private static readonly LocalizableString s_localizableTryGetValueMessage = CreateLocalizableResourceString(nameof(PreferDictionaryTryAddValueMessage));
+        private static readonly LocalizableString s_localizableTryGetValueDescription = CreateLocalizableResourceString(nameof(PreferDictionaryTryAddValueDescription));
+
+        private static readonly DiagnosticDescriptor DiagnosticDescriptor = DiagnosticDescriptorHelper.Create(
+            RuleId,
+            s_localizableTitle,
+            s_localizableTryGetValueMessage,
+            "Performance",
+            RuleLevel.IdeSuggestion,
+            s_localizableTryGetValueDescription,
+            false,
+            false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(DiagnosticDescriptor);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterCompilationStartAction(OnCompilationStart);
+        }
+
+        private static void OnCompilationStart(CompilationStartAnalysisContext context)
+        {
+            if (!TryGetDictionaryTypeAndMembers(context.Compilation, out var iDictionaryType, out var containsKeySymbol, out var addSymbol))
+            {
+                return;
+            }
```

Comment:
I think we should also return if there is no `TryAdd` method.

Reply:
Since `IDictionary` does not have the `TryAdd` method, I don't think this would be the right place to add the check. I could add it in the `IsDictionaryType` method. What do you think?

Reply:
Yes, the idea is that we don't give a warning suggesting to use `TryAdd` where such method doesn't actually exist.